mlflow/entities/run_data.py

```diff
@@ -2,66 +2,69 @@
     Run data (metrics and parameters).
     """
     def __init__(self, metrics=None, params=None, tags=None):
-        self._metrics = []
-        self._params = []
-        self._tags = []
-        if metrics is not None:
-            for m in metrics:
-                self._add_metric(m)
-        if params is not None:
-            for p in params:
-                self._add_param(p)
-        if tags is not None:
-            for t in tags:
-                self._add_tag(t)
+        """
+        Construct a new :py:class:`mlflow.entities.RunData` instance.
+        :param metrics: List of :py:class:`mlflow.entities.Metric`.
+        :param params: List of :py:class:`mlflow.entities.Param`.
+        :param tags: List of :py:class:`mlflow.entities.RunTag`.
+        """
+        # Maintain the original list of metrics so that we can easily convert it back to
+        # protobuf
+        self._metric_objs = metrics or []
+        self._metrics = {metric.key: metric.value for metric in self._metric_objs}
+        self._params = {param.key: param.value for param in (params or [])}
+        self._tags = {tag.key: tag.value for tag in (tags or [])}
 
     @property
     def metrics(self):
-        """List of :py:class:`mlflow.entities.Metric` for the current run."""
+        """
+        Dictionary of string key -> metric value for the current run.
+        For each metric key, the maximum metric value at the maximum timestamp is returned.
+        """
         return self._metrics
 
     @property
     def params(self):
-        """List of :py:class:`mlflow.entities.Param` for the current run."""
+        """Dictionary of param key (string) -> param value for the current run."""
         return self._params
 
     @property
     def tags(self):
-        """List of :py:class:`mlflow.entities.RunTag` for the current run."""
+        """Dictionary of tag key (string) -> tag value for the current run."""
         return self._tags
 
     def _add_metric(self, metric):
-        if isinstance(metric, dict):
-            metric = Metric(metric['key'], metric['value'], metric['timestamp'])
-        self._metrics.append(metric)
+        self._metrics[metric.key] = metric.value
+        self._metric_objs.append(metric)
 
     def _add_param(self, param):
-        if isinstance(param, dict):
-            param = Param(param['key'], param['value'])
-        self._params.append(param)
+        self._params[param.key] = param.value
 
     def _add_tag(self, tag):
-        if isinstance(tag, dict):
-            tag = RunTag(tag['key'], tag['value'])
-        self._tags.append(tag)
+        self._tags[tag.key] = tag.value
 
     def to_proto(self):
         run_data = ProtoRunData()
-        run_data.metrics.extend([m.to_proto() for m in self.metrics])
-        run_data.params.extend([p.to_proto() for p in self.params])
-        run_data.tags.extend([t.to_proto() for t in self.tags])
+        run_data.metrics.extend([m.to_proto() for m in self._metric_objs])
```

Comment:
nit: why extend? why not just assign the value?

Reply:
Apparently protobuf doesn't allow you to assign to `repeated` fields within a protobuf object (from Travis logs):

```
  File "/home/travis/build/mlflow/mlflow/mlflow/server/handlers.py", line 272, in _get_run
response_message.run.MergeFrom(_get_store().get_run(request_message.run_uuid).to_proto())
  File "/home/travis/build/mlflow/mlflow/mlflow/entities/run.py", line 40, in to_proto
    run.data.MergeFrom(self.data.to_proto())
  File "/home/travis/build/mlflow/mlflow/mlflow/entities/run_data.py", line 58, in to_proto
    run_data.metrics = [m.to_proto() for m in self._metric_objs]
AttributeError: Assignment not allowed to repeated field "metrics" in protocol message object.
```